# Week 1–2: Data Cleaning & Exploratory Data Analysis (EDA)

## Project Overview

This notebook is part of my AnalystLab Africa Data Analytics Internship (Batch C).

The objective of this project is to understand, clean, and explore a real-world dataset using Python and Pandas. The analysis aims to improve data quality, identify meaningful patterns, and prepare the dataset for further analysis and visualization.

---

## Learning Objectives

By completing this notebook, I will:

- Understand the dataset structure
- Identify data quality issues
- Clean and prepare the dataset
- Perform exploratory data analysis (EDA)
- Extract business insights
- Document the analysis professionally

---

## Dataset

**Dataset Name:** Online Retail Dataset

This dataset contains transactional records from an online retail store, including invoices, products, quantities, prices, customers, and countries.

---

## Tools Used

- Python
- Pandas
- Jupyter Notebook
- Git & GitHub

# Week 1: Exploratory Data Analysis (EDA)

## AnalystLab Africa Internship

**Intern:** Asres Yelia

---

## 🎯 Objective

In this notebook, I will:
- Load and understand raw datasets
- Perform initial data exploration
- Identify missing values and data quality issues
- Prepare a plan for data cleaning
- Extract early insights

## 🌍 Project Vision

The goal of this internship is to become a job-ready Data Analyst by working on real-world datasets.

Each week builds a key skill:
- Week 1–2: Data Cleaning & EDA
- Week 3: SQL for Analysis
- Week 4: Power BI Dashboards
- Week 5+: Advanced Analytics & Capstone

This notebook represents the first step in building strong analytical thinking and storytelling skills.

## 📥 Loading the Dataset

I will now load the datasets into Pandas to begin exploration.

In [25]:
import pandas as pd

online = pd.read_csv("../data/OnlineRetail.csv", encoding="latin1")
netflix = pd.read_csv("../data/netflix_titles.csv")

## Loading Dataset

While loading the dataset, I encountered an encoding issue.  
This happens when the file is not encoded in UTF-8 format.

To fix this, I specified `encoding='latin1'` while reading the CSV file.

## Data Loading Successful

The dataset was successfully loaded after handling an encoding issue.

This confirms that:
- The file uses a non-UTF-8 encoding format
- `encoding='latin1'` resolves the issue
- Data is now ready for exploration and analysis

## Previewing the Data (.head())

The `.head()` function shows the first 5 rows of the dataset.

This helps me:
- Understand column structure
- See sample values
- Identify data types visually

In [26]:
online.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## Observations from `.head()`

From the preview, I observe:

- Each row represents a single product purchase
- One invoice can contain multiple rows (multiple products per order)
- The dataset includes product, customer, and country information
- Transaction time is stored in `InvoiceDate`
- Customer behavior analysis will be possible using `CustomerID`

In [27]:
online.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


##  Data Quality Assessment

From the `.info()` output, I observed:

###  Missing Values:
- `Description` has some missing values
- `CustomerID` has a significant number of missing values

###  Data Types:
- `InvoiceDate` is stored as object (needs conversion to datetime)
- `CustomerID` is float but should be converted to integer or string

###  Dataset Size:
- 541,909 total records
- 8 columns in total

##  Business Interpretation

- Missing `CustomerID` suggests some transactions are not linked to registered customers
- This may represent guest checkout or incomplete records
- Converting `InvoiceDate` is important for time-based analysis
- Data cleaning will be required before analysis

In [28]:
# Missing values in the dataset
online.isnull().sum()



InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [29]:
# duplicate rows in the dataset
online.duplicated().sum()

np.int64(5268)

##  Data Quality Issues Identified

###  Missing Values:
- `Description`: 1,454 missing values
- `CustomerID`: 135,080 missing values (significant issue)

###  Duplicate Records:
- 5,268 duplicate rows detected


## Interpretation

- Missing `CustomerID` may indicate guest transactions or incomplete customer tracking
- Missing product descriptions may affect product-level analysis
- Duplicate records must be removed to ensure accurate analysis results

## Removing Duplicate Records

Duplicate rows can distort analysis results such as:
- Total sales
- Product popularity
- Customer behavior

Therefore, I will remove duplicate entries from the dataset.

In [30]:
# Remove duplicate rows
online = online.drop_duplicates()

In [31]:
# checking again for duplicate rows
online.duplicated().sum()

np.int64(0)

##  Duplicate Check After Cleaning

After removing duplicates, I re-checked the dataset.

Result:
- No duplicate rows found (0 duplicates)

This confirms that the data cleaning step was successful.

## Handling Missing Values

The dataset contains missing values in:
- Description
- CustomerID

These must be handled before analysis.

In [39]:
# Remove rows with missing values containing CustomerID
online = online.dropna(subset=["CustomerID"])


In [ ]:
# Remove rows with missing values

online = online.dropna(subset=["Description"])
# print the number of rows and columns in the dataset
print("Number of rows:", online.col

Number of rows: 401604
Number of columns: 8
